crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -1702,10 +1697,9 @@ where
                 ..
             } => {
                 if self.settings.rules.enabled(Rule::UnusedLoopControlVariable) {
-                    self.deferred.for_loops.push((
-                        stmt,
-                        (self.ctx.scope_stack.clone(), self.ctx.parents.clone()),
-                    ));
+                    self.deferred
+                        .for_loops
+                        .push((stmt, (self.ctx.scope_id, self.ctx.parents.clone())));
```

Comment:
Is it still necessary to clone `parents` or can we even remove `parents`?

Reply:
I think we can remove `parents` soon, but it requires a similar approach. We do need to set the list of parents when we "restore" state, but we should be able to use IDs with parent pointers, just like with scopes.

Reply:
We may also want to consider using `indextree` for statements, to enable us to traverse up and get symbolic representations?

Reply:
(Ignore the `indextree` reference here, I think we can make improvements to parent tracking but that crate might end up being irrelevant.)